go/pkg/pass2/check-acl.1.md
# check-acl 1 "" Netspoc "User Manual"

# NAME

check-acl - Check if packets would pass given access list

# SYNOPSIS

check-acl [-f file] code/ROUTER ACL-NAME [PACKET-DESCRIPTION ...]

# DESCRIPTION

This command checks if given packets would be permitted or denied
by specified ACL.
ACL is read from code file that was generated by Netspoc for given router.
Packet descriptions are given on command line or read from file.
Each packet description is written to STDOUT,
prefixed with:

- `permit` if packet would be permitted
- `deny` if packet would be denied.

A packet description consists of 4 blank separated fields:

1. source IP address
2. destination IP address
3. protocol: `tcp` | `udp` | `icmp` | `proto`
4. port number | icmp-type/code | protcol number

# OPTIONS

**-f**, **--file**
:   Read packet descriptions from file.

**-h**, **--help**
:   Print a brief help message and exit.

# EXAMPLES

With this Netspoc configuration:

    ## START
    network:n1 = { ip = 10.1.1.0/24; }
    network:n2 = { ip = 10.1.2.0/24; }

    router:r1 = {
     managed;
     model = ASA;
     interface:n1 = { ip = 10.1.1.1; hardware = n1; }
     interface:n2 = { ip = 10.1.2.1; hardware = n2; }
    }

    service:s1 = {
     user = network:n1;
     permit src = user; dst = network:n2; prt = tcp 80-90, icmp 3/3;
    }
    ## END

and this file with packet descriptions:

    10.1.1.11 10.1.2.12 tcp 85
    10.1.1.11 10.0.0.0 udp 123
    10.1.1.11 10.1.2.12 icmp 3/3
    10.1.1.11 10.1.2.12 icmp 3/13

we would get this result:

`check-acl -f file code/r1 n1_in`

    permit 10.1.1.11 10.1.2.12 tcp 85
    deny   10.1.1.11 10.0.0.0 udp 123
    permit 10.1.1.11 10.1.2.12 icmp 3/3
    deny   10.1.1.11 10.1.2.12 icmp 3/13

# COPYRIGHT AND DISCLAIMER

(c) 2025 by Heinz Knutzen, heinz.knutzen@googlemail.com

http://hknutzen.github.com/Netspoc

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


Changes
{{$NEXT}}

 - export-netvis: export Dualstack-Address of network
   in field Address2 of node.

6.079     2025-03-17 12:53:54+01:00 Europe/Berlin

 - If service has some owner with name "DA_*",
   this service is marked as visible for all owners with name "DA_*"
   in exported data for Netspoc-Web.
 - Command 'print-path' now accepts dual stack objects.

6.078     2025-02-24 14:42:54+01:00 Europe/Berlin

 - New commandline programs export-netvis and print-path have been added.

    These tools output information as JSON to be used in
    a graph visualization.

    export-netvis creates a list of all networks and routers
    of the given topology with information useful for visualizing
    the topology.

    print-path analyzes the given topology and creates a list of
    names of network nodes (routers and networks) between source and
    destination passed via command line.
    To create the list, print-path uses the same path-walk algorithm
    as used in netspoc main program to create ruleset for devices.

 - Aggregate with IPv4 IP 0.0.0.0/0 no longer matches IPv6 addresses.
   Aggregate with IPv6 IP ::/0 no longer matches IPv4 addresses.
 - Fixed split service from auto intf with IPv4, IPv6
 - Command 'export-netspoc' no longer strips comments and ";"
   from description in details of exported service.

6.077     2025-01-27 11:24:26+01:00 Europe/Berlin

 - Allow dual stack service to have protocols icmp and icmpv6 together.
   Generated rules automatically will use icmp for IPv4 and icmpv6 for IPv6.
 - Removed support for separate IPv6 topology in subdirectory 'ipv6/'.
 - Removed options '--ipv6', '-6'.
 - Removed support for NAT of IPv6 objects.
   After changing attributes of IPv6 addresses from 'ip' to 'ip6',
   supporting NAT for IPv6 would need to change NAT attributes
   to 'ip6', 'hidden6', 'identity6' as well.
   This seems to be superflous, since nobody uses NAT for IPv6.
 - Fixed duplicate hosts in print-group from wrong IPv4 network
   in IPv6 aggregate.
 - Fixed panic in print-group from wrong IPv4 network in IPv6 hosts
   generated by 'auto_ipv6_hosts'.
 - Fixed panic in export-netspoc with rule from IPv4 network to
   combined IPv4/v6 auto interface.
 - Changed 'format-netspoc' to print attributes "ip" and "ip6" in
   one line, if object has no more than 3 attributes.

6.076     2025-01-10 09:39:54+01:00 Europe/Berlin

 - Fixed dual stack bridge. IPv4 and IPv6 interfaces were accidently
   processed together.

6.075     2024-12-19 11:53:37+01:00 Europe/Berlin

 - Fix: Attribute 'policy_distribution_point' at dual stack router
   is only used for matching IPv4/IPv6 part of this router.
 - Fixed command 'export-netspoc'.
   Show IPv4 areas for combined zone. Must not accidently overwrite
   IPv4 areas by IPv6 areas, when combined zones have same name.

6.074     2024-12-16 11:52:23+01:00 Europe/Berlin

 - Fixed wrong error on duplicate interface from router in ipv6/ directory.

6.073     2024-12-16 10:16:33+01:00 Europe/Berlin

 - Added support for dual stack objects, having both, IPv4 and IPv6 addresses.
   This simplifies the modeling of a dual stack toplology.
   It is no longer neccessary to model a separate IPv4 and IPv6 topology.
   Rules between dual stack objects will generate ACls for IPv4 and IPv6.
   Rules between dual stack object and pure IPv4 object will silently ignore
   IPv6 address and generate only ACL for IPv4.
   New attributes have been introduced to define dual stack objects:
   - 'ip6' at network, host, interface and aggregate.
   - 'range6' at host.
   - 'unnumbered6' at network and interface.
   - 'negotiated6' at interface.
   Other changes resulting from use of dual stack objects:
   - If dual stack objects are used as border or inclusive border of an area,
     this defines two areas with identical name:
     one in IPv4 topology and one in IPv6 topology.
   - If dual stack objects are used as interfaces of an pathrestriction,
     this also defines two pathrestrictions in IPv4 and IPv6 topology.
     If the second pathrestriction has only one interface or only interfaces
     outside of a loop, it is silently ignored.
   - New attributes 'ipv4_only' and 'ipv6_only' may be used at service or area.
     This will enable only IPv4 or IPv6 part of dual stack objects.
   - The following attributes are applied only to IPv4 part
     if used in dual stack objects:
     'nat', 'bind_nat', 'subnet_of, 'hub', 'spoke'.
   - Command 'export-netspoc' has been changed to show IPv6 address of
     dual stack objects in attribute 'ip6'.
 - Added new attribute 'auto_ipv6_hosts' to automatically generate
   dual stack hosts from pure IPv4 hosts.
   It will generate IPv6 addresses for hosts
   by combining its IPv4 adress with the IPv6 address of its network.
   These attribute values are provided:
   - auto_ipv6_hosts = readable
     network: ip6 = 2001:db8:1:1::/64;
     hosts: ip = 172.17.1.48;
     ==> 2001:db8:1:1:172:17:1:48
   - auto_ipv6_hosts = binary
     network: ip6 = 2001:db8:1:1::/64;
     hosts: ip = 172.17.1.48;
     ==> 2001:db8:1:1::ac11:130
   - auto_ipv6_hosts = none
     No IPv6 address is generated
   This attribute can be used at network, area and host.
 - Fixed command 'cut-netspoc':
   Full path to management_instance is marked now.

6.072     2024-09-05 10:22:13+02:00 Europe/Berlin

 - Fixed command 'remove-from-netspoc'.
   A service is no longer removed accidently, if only some rules of
   a service become empty, but other rules remain.
 - Message "swap objects of 'user' and objects of rules" is shown
   even if no attribute multi_owner is given.
 - Fixed check for NAT network being subnet of some other network.
   If multiple networks have dynamic NAT to same address,
   now a warning is given for each network.
 - Fixed check for useless attribute 'subnet_of'
   where supernet has attribute 'has_subnets'.

6.071     2024-07-29 16:47:58+02:00 Europe/Berlin

 - Fixed generated ACL at router R with managed=local for rule where

go/pkg/pass2/check-acl.go
package pass2

import (
	"fmt"
	"io"
	"net"
	"os"
	"slices"
	"strconv"
	"strings"

	"github.com/hknutzen/Netspoc/go/pkg/fileop"
	"github.com/hknutzen/Netspoc/go/pkg/oslink"
	"github.com/spf13/pflag"
)

type packet struct {
	src string
	dst string
	prt string
}

func CheckACLMain(d oslink.Data) int {
	fs := pflag.NewFlagSet(d.Args[0], pflag.ContinueOnError)

	// Setup custom usage function.
	fs.Usage = func() {
		fmt.Fprintf(d.Stderr,
			"Usage: %s [-f file] code/router ACL-NAME ['ip1 ip2 tcp|udp port']...\n%s",
			d.Args[0], fs.FlagUsages())
	}

	// Command line flags
	fromFile := fs.StringP("file", "f", "", "Read packet descriptions from file")
	if err := fs.Parse(d.Args[1:]); err != nil {
		if err == pflag.ErrHelp {
			return 1
		}
		fmt.Fprintf(d.Stderr, "Error: %s\n", err)
		fs.Usage()
		return 1
	}

	// Initialize packet descriptions.
	var packets []*packet
	if *fromFile != "" {
		var err error
		packets, err = readPackets(d.Stderr, *fromFile)
		if err != nil {
			fmt.Fprintf(d.Stderr, "Error: %s\n", err)
			return 1
		}
	}

	// Argument processing
	args := fs.Args()
	if !(*fromFile != "" && len(args) >= 2 || len(args) >= 3) {
		fs.Usage()
		return 1
	}
	path := args[0]
	if !strings.HasSuffix(path, ".rules") {
		path += ".rules"
	}
	if !fileop.IsRegular(path) {
		fmt.Fprintf(d.Stderr, "Error: Can't find file %s\n", path)
		return 1
	}

	acl := args[1]

	packets = append(packets, parsePackets(d.Stderr, args[2:])...)

	// Check, which packets match ACL.
	return checkACL(d, path, acl, packets)
}

func readPackets(stderr io.Writer, path string) ([]*packet, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("Can't %s", err)
	}
	lines := strings.Split(string(data), "\n")
	return parsePackets(stderr, lines), nil
}

func parsePackets(stderr io.Writer, lines []string) []*packet {
	var result []*packet
	warn := func(format string, args ...any) {
		fmt.Fprintf(stderr, "Warning: "+format+"\n", args...)
	}
	checkIP := func(s string) string {
		ip := net.ParseIP(s)
		if ip == nil {
			warn("Ignored packet with invalid IP address: %s", s)
			return ""
		}
		return ip.String()
	}
	checkNum := func(s string, max int) string {
		num, err := strconv.Atoi(s)
		if err != nil || num < 0 || num >= max {
			warn("Ignored packet with invalid protocol number: %s", s)
			return ""
		}
		return strconv.Itoa(num)
	}
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		line = strings.ToLower(line)
		fields := strings.Fields(line)
		if len(fields) != 4 {
			warn("Ignored packet, must have exactly 4 words: %s", line)
			continue
		}
		ip1 := fields[0]
		ip2 := fields[1]
		prt := fields[2]
		ext := fields[3]
		ip1 = checkIP(ip1)
		ip2 = checkIP(ip2)
		if ip1 == "" || ip2 == "" {
			continue
		}
		switch prt {
		case "tcp", "udp":
			ext = checkNum(ext, 65536)
		case "icmp":
			typ, code, found := strings.Cut(ext, "/")
			if found {
				typ = checkNum(typ, 256)
				code = checkNum(code, 256)
				if typ == "" || code == "" {
					continue
				}
				ext = typ + "/" + code
			} else {
				warn("Ignored icmp packet with invalid type/code: %s", line)
				continue
			}
		case "proto":
			ext = checkNum(ext, 256)
		default:
			warn("Ignored packet with unexpected protocol: %s", line)
			continue
		}
		if ext == "" {
			continue
		}
		p := packet{src: ip1, dst: ip2, prt: prt + " " + ext}
		result = append(result, &p)
	}
	return result
}

// Print each packet to STDOUT.
// Packet, that matches ACL is prefixed with "permit".
// Other packet is prefixed with "deny  ".
func checkACL(d oslink.Data,
	path, acl string, packets []*packet) int {

	rData := readJSON(path)
	var aInfo *aclInfo
	for _, a := range rData.acls {
		if a.name == acl {
			aInfo = a
			break
		}
	}
	if aInfo == nil {
		fmt.Fprintf(d.Stderr, "Error: Unknown ACL: %s\n", acl)
		return 1
	}
	// Remember number of original rules.
	sz := len(aInfo.rules)
	// Add packets as rules.
	addPackets(aInfo, packets)
	// Set relation between original and added rules.
	setupPrtRelation(aInfo.prt2obj)
	setupIPNetRelation(aInfo.ipNet2obj)
	// Remember added original rules because aInfo gets changed
	// during optimization.
	orig := slices.Clone(aInfo.rules[sz:])
	// Optimize rules; marks duplicate and redundant rules as '.deleted'.
	optimizeRules(aInfo)
	// Check added rules.
	for _, p := range orig {
		var action string
		if p.deleted {
			// Packet was redundant to original rule and hence can pass.
			action = "permit"
		} else {
			action = "deny  "
		}
		ip1 := p.src.Addr().String()
		ip2 := p.dst.Addr().String()
		prt := p.prt.name
		fmt.Fprintf(d.Stdout, "%s %s %s %s\n", action, ip1, ip2, prt)
	}
	return 0
}

func addPackets(a *aclInfo, l []*packet) {
	rules := a.rules
	ipNet2obj := a.ipNet2obj
	prt2obj := a.prt2obj
	seen := make(map[packet]bool)
	for _, p := range l {
		// Ignore duplicate packets, because duplicates are redundant to
		// each other and would therefore show inconsistent results.
		if seen[*p] {
			continue
		}
		seen[*p] = true
		ipObj := func(s string) *ipNet {
			if strings.Contains(s, ":") {
				s += "/128"
			} else {
				s += "/32"
			}
			return getIPNet(s, ipNet2obj)
		}
		src := ipObj(p.src)
		dst := ipObj(p.dst)
		prt := getPrtObj(p.prt, prt2obj)
		rules.push(&ciscoRule{src: src, dst: dst, prt: prt})
	}
	a.rules = rules
}
